### PR TITLE
Add test cases for SYN-514 and SYN-515

### PR DESCRIPTION
**Pull request by richvdh**
A couple of edge-cases in v2 /sync which don't look quite right to me.

### CODE REVIEW

File: tests/31sync/06state.pl [hunk 1]
```diff
@@ -338,6 +338,64 @@
...
+
+      my ( $user, $filter_id, $room_id, $next_batch );
+
+      my $filter = {
+         room => {
+            timeline  => { types => [] },
+            state     => { types => [ "a.madeup.test.state" ] },
+            ephemeral => { types => [] },
+         },
+         presence => {types => [] },
```
Comment (leonerd) [inline tests/31sync/06state.pl]: Wants whitespace before `types`

File: tests/31sync/06state.pl [hunk 2]
```diff
@@ -338,6 +338,64 @@
...
+      matrix_register_user_with_filter( $http, $filter )->then( sub {
+         ( $user, $filter_id ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_put_room_state( $user, $room_id,
+            type      => "a.madeup.test.state",
+            content   => { "my_key" => 1 },
```
Comment (leonerd) [inline tests/31sync/06state.pl]: No need to quote `"my_key"`

File: tests/31sync/06state.pl [hunk 3]
```diff
@@ -338,6 +338,64 @@
...
+            state_key => ""
+         );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         $next_batch = $body->{next_batch};
```
Comment (leonerd) [inline tests/31sync/06state.pl]: You're setting `$next_batch` but then don't appear to be actually using it afterwards

File: tests/31sync/06state.pl [hunk 4]
```diff
@@ -403,3 +461,69 @@
...
+      })->then( sub {
+         matrix_leave_room( $user_b, $room_id );
+      })->then( sub {
+         matrix_sync( $user_b, filter => $filter_id_b, since => $next_b );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{archived}{$room_id};
+         require_json_keys( $room, qw( event_map timeline state ephemeral ));
+         @{ $room->{state}{events} } == 1
```
Comment (leonerd) [inline tests/31sync/06state.pl]: This might be a bit neater as:

    my $eventcount = scalar @{ $room->{state}{events} };
    $eventcount == 1 or
       die "Expected one state event, got $eventcount";

**Comment by leonerd** [comment]
Assigning to @NegativeMjark to comment on the semantic meaning of the v2 sync tests

File: tests/31sync/06state.pl [hunk 5]
```diff
@@ -338,6 +338,64 @@
...
+            or die "Expected one state event";
+
+         matrix_put_room_state( $user, $room_id,
+            type      => "m.room.member",
+            content   => { "membership" => "join",
+                           "displayname" => "boris" },
+            state_key => $user->user_id,
+         );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id );
```
Comment (NegativeMjark) [inline tests/31sync/06state.pl]: Shouldn't this have a ``since => $next_batch``?

File: tests/31sync/06state.pl (hunk 2 above)
Comment (richvdh) [inline tests/31sync/06state.pl]: true, but it is consistent with the rest of the file...

File: tests/31sync/06state.pl [hunk 6]
```diff
@@ -338,6 +338,64 @@
...
+      })->then( sub {
+         my ( $body ) = @_;
+
+         @{ $body->{rooms}{joined}{$room_id}{state}{events} } == 0
```
Comment (NegativeMjark) [inline tests/31sync/06state.pl]: Aren't we expecting to see the display change state event? Or does this PR assume that the state change will only be in the timeline?

File: tests/31sync/06state.pl (hunk 1 above)
Comment (richvdh) [inline tests/31sync/06state.pl]: Done.

File: tests/31sync/06state.pl (hunk 3 above)
Comment (richvdh) [inline tests/31sync/06state.pl]: Ooops.

File: tests/31sync/06state.pl (hunk 4 above)
Comment (richvdh) [inline tests/31sync/06state.pl]: Done. thanks.

File: tests/31sync/06state.pl (hunk 5 above)
Comment (richvdh) [inline tests/31sync/06state.pl]: indeed.

File: tests/31sync/06state.pl (hunk 6 above)
Comment (richvdh) [inline tests/31sync/06state.pl]: I think I just messed up, to be honest. I've made it tolerate zero or one events, so that it will work whether or not SPEC-254 is finished.

Comment (richvdh) [inline tests/31sync/06state.pl]: Wait, I'm being a crank. The m.room.member event is removed by the filter.

**Comment by richvdh** [comment]
Thanks both for your feedback. Anybody want to doublecheck?

File: tests/31sync/06state.pl (hunk 2 above)
Comment (leonerd) [inline tests/31sync/06state.pl]: Fair enough; consistency FTW

**Comment by leonerd** [comment]
LGTM

**Comment by NegativeMjark** [comment]
LGTM
